### DISCUSSION TITLE
Handle based pagination

### DISCUSSION THREAD

**Discussion by xuala69** [simolus3/drift#2091] [Q&A] [answered]
Hello,

I'm trying to implement pagination using handle (example..get 15 items after item with this id/handle). Can this be done using the currently existing dart queries or I'll have to write a custom SQL query manually?

Thanks

**Comment by madsane29** [comment, marked as answer]
Guess the author will give you a better answer but maybe playing with `limit()` would make the magic?

```
  Future<List<*Entity*>> getAllWithLimitAndOffset(int limit, int? offset) async {
        return (select(*tablename*)
          ..orderBy(
            ([
              (t) => OrderingTerm(expression: t.id, mode: OrderingMode.asc), // put this here so surely after every call the results will be in the same order
            ]),
          )
          ..limit(limit, offset: offset))
        .get();
  }
```

Where the `int limit` would be how many you want to show (let's say 10) and the `int? offset` is how many rows you wanna skip.

Then in code it would be something like:

```
List<*Entity*> list = [];

list = await db.getAllWithLimitAndOffset(10, currentPage * numOfRowsNeeded - numOfRowsNeeded);
```

Let's say you have 95 rows with ids 1 to 95.

First you call the `getAllWithLimitAndOffset` with parameters of (10, 1 * 10 - 10) -> offset == 0 -> get the results with id 1-10.
Second call with parmeters of (10, 2 * 10 - 10) -> offset == 10 -> get the results with id 11-20
Third call with parameters of (10, 3 * 10 - 10) -> offset == 20 -> get the results with id 21-30

**Reply by xuala69** [reply]
This would work if the id for the table is integer but I have to use a string for the id as the server stores it as an uuid.So i can't use limit and offset with where to jump to any data after a specific id. I've tried sorting my data descending order of the createdAt time and tried to get data that is after the given time and it seems to work. I'm hoping there's a better way

**Reply by madsane29** [reply]
Even if it is a string, if it is an id (== unique), it won't sort differently each call.

**Reply by simolus3** [reply]
Yes exactly. In sqlite3, the ordering of results is technically undefined unless you use an explicit `order by` clause, so I recommend adding one if the order is important. Of course, you can also order by a string key, and `limit` should work with that.

**Reply by xuala69** [reply]
I did not know that. Thanks for clearing it all up for me guys.
